nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/dto/DtoFactory.java

```diff
@@ -1083,13 +1083,15 @@ public ProcessorStatusDTO createProcessorStatusDto(final ProcessorStatus procSta
         dto.setGroupId(procStatus.getGroupId());
         dto.setName(procStatus.getName());
         dto.setStatsLastRefreshed(new Date());
+        dto.setRunStatus(procStatus.getRunStatus().toString());
```

Comment:
In addition to setting this here, we need to update it when merging responses in a clustered scenario. This merging happens to account for cases when a processor is invalid based on an environmental condition. A good example of this is a processor configured with a directory where that directory does not exist on one node of a cluster. On that node, the processor is invalid, but on the others that processor would be considered stopped. In this case, we need to ensure the invalid run status takes precedence. You can see an example of this here [1].

I think the easiest way to handle this would be to update the runStatus in the ProcessorStatusDTO when the method referenced in [1] returns when appropriate. Let me know if you have any questions about this.

[1] https://github.com/apache/nifi/blob/2da72f15243ef62d47369e8a87cc202a82dbe004/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/manager/StatusMerger.java#L403

Reply:
So, something like target.setRunStatus(toMerge.getRunStatus()); ?

Reply:
You cannot just set target.setRunStatus(toMerge.getRunStatus) in case target is invalid. I believe you have two options...

1) Duplicate the isValid check

        if (RunStatus.Invalid.name().equals(toMerge.getRunStatus())) {
            target.setRunStatus(RunStatus.Invalid.name());
        }

2) Copy the calculated runStatus from the aggregate snapshot

        target.setRunStatus(target.getAggregateSnapshot().getRunStatus());

Reply:
I will add a "else" to the already existing check and set the status in that case. I believe, it will be find.

Reply:
I'm not sure I follow what "else" you're referring to. No change is needed when merging snapshots (line 384), that part is already correct. What we need to account for is the runStatus when merging status (line 354). We can account for the runStatus there using one of the two options I proposed above after merging the aggregate snapshot on line 363.